fml/closure.h

```diff
@@ -7,10 +7,50 @@
 
 #include <functional>
 
+#include "flutter/fml/macros.h"
+
 namespace fml {
 
 using closure = std::function<void()>;
 
+//------------------------------------------------------------------------------
+/// @brief      Wraps a closure that is invoked in the destructor unless
+///             released by the caller.
+///
+///             This is especially useful in dealing with APIs that return a
+///             resource by accepting ownership of a sub-resource and a closure
+///             that releases that resource. When such APIs are chained, each
+///             link in the chain must check that the next member in the chain
+///             has accepted the resource. If not, it must invoke the closure
+///             eagerly. Not doing this results in a resource leak in the
+///             erroneous case. Using this wrapper, the closure can be released
+///             once the next call in the chain has successfully accepted
+///             ownership of the resource. If not, the closure gets invoked
+///             automatically at the end of the scope. This covers the cases
+///             where there are early returns as well.
+///
+class AutoFireClosure {
```

Comment:
I already introduced a resource leak on error as I was refactoring the initial version of the implementation. Since it is so easy to introduce such resource leaks in C and C++ interop code (in error cases no less), I'd rather make this a pattern we follow.

About the name, maybe `ScopedFireClosure` is more immediately descriptive?

Reply:
I like `Scoped`, since `Scoped*` is a common naming pattern. What about `ScopedCleanupClosure`, which more clearly indicates that it fires on destruction?

Reply:
`ScopedCleanupClosure` sgtm

Reply:
Done.